File: runtime/apps/gui/sdk/gateway/src/types/agent.ts
export type Agent = {
  name: string;
  description: string;
  mode: string;
  native: boolean;
  hidden: boolean;
  model?: {
    providerID: string;
    modelID: string;
  } | null;
  options: Record<string, unknown>;
  permission: {
    allow: string[];
    deny: string[];
  };
};

export type AgentSummary = {
  id: string;
  name: string;
  description: string;
  source: "dynamic" | "static";
  path: string;
  aliases: string[];
  capabilities: string[];
  provider?: string | null;
  hidden: boolean;
};

export type AgentConfig = {
  agent_name: string;
  description?: string | null;
  aliases?: string[];
  icon_emoji?: string | null;
  agent_directory?: string;
  parent_agent_id?: string | null;
  report_to_user?: boolean;
  default_config?: boolean;
  provider?: unknown;
  agent_prompt?: unknown[];
  agent_capabilities?: unknown[];
  validator?: unknown;
  avatar?: AgentAvatarConfig;
};

export type AgentAvatarConfig = {
  persona_id?: string;
  role?: string;
  display_mode?: "hidden" | "static" | "dynamic";
  pixel_size: number;
  threshold: number;
};

export type StoredAgent = {
  summary: AgentSummary;
  config: AgentConfig;
  prompt?: string | null;
};

export type AgentUpsertRequest = {
  id?: string;
  config?: AgentConfig;
  prompt?: string;
};

File: runtime/crates/gateway/src/contracts/agent.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Agent {
    pub name: String,
    pub description: String,
    pub mode: String,
    pub native: bool,
    pub hidden: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<AgentModel>,
    #[serde(default)]
    pub options: HashMap<String, serde_json::Value>,
    pub permission: PermissionRuleset,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct UpsertAgentRequest {
    #[serde(default)]
    pub id: Option<String>,
    #[serde(default)]
    pub config: Option<tura_agents::store::AgentConfig>,
    #[serde(default)]
    pub prompt: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentModel {
    #[serde(rename = "providerID")]
    pub provider_id: String,
    #[serde(rename = "modelID")]
    pub model_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionRuleset {
    pub allow: Vec<String>,
    pub deny: Vec<String>,
}
